sql/catalog,schemachanger: allow UDFs in ON UPDATE and computed expressions

rafiss opened the pull request:
We already had the dependency tracking in place, but there was still a
block in place since we don't yet support UDFs in backfills. There were
just a few places relating to ON UPDATE expressions that needed to be
updated to emit the mutation event that modifies dependency back
references.

This patch removes the block within the schemachanger, so that now the
only thing that gets blocked is actually attempting to use a UDF in the
context of a backfill.

informs #87699
fixes #115346
Release note (sql change): Computed column expressions and ON UPDATE
expressions can now reference user-defined functions.

rafiss committed:
sql/catalog,schemachanger: allow UDFs in ON UPDATE and computed expressions

We already had the dependency tracking in place, but there was still a
block in place since we don't yet support UDFs in backfills. There were
just a few places relating to ON UPDATE expressions that needed to be
updated to emit the mutation event that modifies dependency back
references.

This patch removes the block within the schemachanger, so that now the
only thing that gets blocked is actually attempting to use a UDF in the
context of a backfill.

Release note (sql change): Computed column expressions and ON UPDATE
expressions can now reference user-defined functions.

rafiss committed:
logictest: add tests for using UDF from an index

Release note: None

rafiss committed:
logictest: update tests to unpack tuples

Release note: None

fqazi reviewed (changes requested):
> Reviewed 27 of 27 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147817)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/sql/alter_table.go` line 1401 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147817#-OS05iISERBPN-9xO8JH:-OS05iISERBPN-9xO8JI:b-kdq2v4) ([raw file](https://github.com/cockroachdb/cockroach/blob/44563fe2bdd69b081ce60132ee48cc50fd0b4067/pkg/sql/alter_table.go#L1401)):*
> ```Go
> 	}
> 
> 	if err := funcdesc.MaybeFailOnUDFUsage(typedExpr, context, p.EvalContext().Settings.Version.ActiveVersionOrEmpty(p.ctx)); err != nil {
> ```

Because we have changes in scmutationexec, we probably want to version gate the dropping of this check? I think a mixed version cluster may not correctly update references if the plan ever ends up the job phase.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
Just one minor concern, but over all this code looks good.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147817#-:-OS06nGa6GaJjA5LqHRW:bnsi1u0)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)



<!-- Sent from Reviewable.io -->

rafiss committed:
schemachanger: track partial index dependencies on functions

This commit adds the plumbing necessary to find referenced UDF IDs in a
partial index predicate, and updates the declarative schema changer to
add back-references.

Release note (sql change): Partial indexes can now reference
user-defined function.

rafiss committed:
sql: support index->UDF dependency tracking in legacy schema changer

This matches the dependency tracking behavior that the declarative
schema changer has.

Release note: None

rafiss committed:
sql/catalog: disallow UDF references in mixed version state

Since the support for UDF dependency tracking required a new schema
change plan, we need to block this in a mixed version state.

Release note: None

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147817)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/alter_table.go` line 1401 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147817#-OS05iISERBPN-9xO8JH:-OSQA-s8ER4ANUXL9jLC:b-8ie7ac) ([raw file](https://github.com/cockroachdb/cockroach/blob/44563fe2bdd69b081ce60132ee48cc50fd0b4067/pkg/sql/alter_table.go#L1401)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Because we have changes in scmutationexec, we probably want to version gate the dropping of this check? I think a mixed version cluster may not correctly update references if the plan ever ends up the job phase.
</blockquote></details>

done; i brought back the mixed version checks, and also added support in legacy schema changer.


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

One minor question for the reference clean up logic in the declarative schema changer

> Reviewed 13 of 13 files at r2, 2 of 2 files at r3, 7 of 9 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147817#-:-OSQGanS9LNLaLrg1nOp:bqng4z6)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained

___
*[`pkg/sql/schemachanger/scexec/scmutationexec/references.go` line 442 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147817#-OSQGKfGCCOfUzFxgg-k:-OSQGKfGCCOfUzFxgg-l:bcg7bdi) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5008431f1a15ff1e09aa68e1f03537a16e5e4ef/pkg/sql/schemachanger/scexec/scmutationexec/references.go#L442)):*
> ```Go
> 	}
> 	var fnIDsInUse catalog.DescriptorIDSet
> 	if !tblDesc.Dropped() {
> ```

What cleans up back references in functions if we are dropping an object?


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147817)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/schemachanger/scexec/scmutationexec/references.go` line 442 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147817#-OSQGKfGCCOfUzFxgg-k:-OSQOHrcA_S_itziP-fd:b-mx21xo) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5008431f1a15ff1e09aa68e1f03537a16e5e4ef/pkg/sql/schemachanger/scexec/scmutationexec/references.go#L442)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

What cleans up back references in functions if we are dropping an object?
</blockquote></details>

this function is for back-refs to indexes, which only applies for partial index predicates. this mutation is emitted in opgen_secondary_index.go.

back-refs to columns (which include index expressions that use a UDF) are cleaned up in `RemoveTableIndexBackReferencesInFunctions`, which was a pre-existing funciton. that mutation is emitted in the `default`, `on update` and `computed column` ops.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147817)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @fqazi)

___
*[`pkg/sql/schemachanger/scexec/scmutationexec/references.go` line 442 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147817#-OSQGKfGCCOfUzFxgg-k:-OSQOsgVAypa2YUkq32k:b-mx2ez0) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5008431f1a15ff1e09aa68e1f03537a16e5e4ef/pkg/sql/schemachanger/scexec/scmutationexec/references.go#L442)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this function is for back-refs to indexes, which only applies for partial index predicates. this mutation is emitted in opgen_secondary_index.go.

back-refs to columns (which include index expressions that use a UDF) are cleaned up in `RemoveTableIndexBackReferencesInFunctions`, which was a pre-existing funciton. that mutation is emitted in the `default`, `on update` and `computed column` ops.
</blockquote></details>

agh, i meant to write `RemoveTableColumnBackReferencesInFunctions` in that second paragraph.


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147817)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained

___
*[`pkg/sql/schemachanger/scexec/scmutationexec/references.go` line 442 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/147817#-OSQGKfGCCOfUzFxgg-k:-OSQOy4IApetBYuOELDD:b-m43kg9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b5008431f1a15ff1e09aa68e1f03537a16e5e4ef/pkg/sql/schemachanger/scexec/scmutationexec/references.go#L442)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

agh, i meant to write `RemoveTableColumnBackReferencesInFunctions` in that second paragraph.
</blockquote></details>

Done.

Ah perfect.


<!-- Sent from Reviewable.io -->

rafiss commented:
tftr!


bors r+
